### PR TITLE
Default fallback for project/department/car on invoice

### LINKED ISSUE #358: Automatically select default project - journal.edit
Bookkeeping to an account with default project should select the project automatically.

Example: If you download an invoice where supplier has default result account 4001. And result account 4001 has default project 123. When you bookkeep the invoice, project 123 should be selected since it's default for account 4001.
Same goes for department and car, unless it(project/department/car) is already selected(downloaded from fb).

### PR DESCRIPTION
**Pull request by mladjo2505**
Closes #358

### CODE REVIEW

File: modules/invoicein/model/invoicein.class.php [hunk 1]
```diff
@@ -400,11 +400,36 @@ public function journal() {
...
+                        $VoucherH['voucher_DepartmentID'] = $InvoiceO->Department;
+                    } else {
+                        $VoucherH['voucher_DepartmentID'] = $invoice_supplier_motkonto_default_department;
+                    }
+                    if($InvoiceO->Project) {
+                        $VoucherH['voucher_ProjectID'] = $InvoiceO->Project;
+                    } else {
+                        $VoucherH['voucher_ProjectID'] = $invoice_supplier_motkonto_default_project;
+                    }
+
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php]: Only assign default directly to the VoucherH['voucher_Dep/Proj'] if not set in $invoice->Dep/Proj

File: modules/invoicein/model/invoicein.class.php [hunk 2]
```diff
@@ -479,6 +504,18 @@ public function journal() {
...
+                        if($line->IsOnlyTax) {
+                            $VoucherH['voucher_AccountPlanID'] = 2710;
+                        }
+                        $line_accountplan = accounting::get_accountplan_object($VoucherH['voucher_AccountPlanID']);
+                        if (!empty($line_accountplan) && $line_accountplan->EnableCar == 1 && !is_null($line_accountplan->CarID)) {
+                          $accountplan_default_car = $line_accountplan->CarID;
+                        } else {
+                          $accountplan_default_car = NULL;
+                        }
+
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php]: Same as above

**Comment by mladjo2505** [comment]
Should be redone after discussion with Arnt.

File: modules/invoicein/model/invoicein.class.php (hunk 2 above)
Comment (mladjo2505) [reply modules/invoicein/model/invoicein.class.php]: Not relevant anymore

**Comment by mladjo2505** [comment]
We should rework EHF part for InvoiceLine->AccountingCost to send and receive Project/Department/Car on both FB and LODO

**Comment by mladjo2505** [comment]
Fixed but not tested enough.
I would like to test some more.

File: modules/fakturabank/model/fakturabank.class.php [hunk 3]
```diff
@@ -1000,21 +1000,44 @@ private function validate_incoming($invoicesO) {
...
 
-                  if ($line->Item->AdditionalItemProperty) {
-                    if(!is_array($line->Item->AdditionalItemProperty)) {
-                      $line->Item->AdditionalItemProperty = array($line->Item->AdditionalItemProperty);
-                    }
-                    includelogic("car/car");
-                    foreach($line->Item->AdditionalItemProperty as $additional_item_property) {
-                      if ($additional_item_property->Name == 'Car') {
-                        $query = "select * from car where CarCode='" . $additional_item_property->Value . "' and ". car::car_active_sql("car.CarID", $InvoiceO->IssueDate) ."=1";
+                  includelogic("car/car");
```
Comment (mladjo2505) [inline modules/fakturabank/model/fakturabank.class.php]: Should do this for outgoing as well.

File: modules/invoice/model/invoice.class.php [hunk 4]
```diff
@@ -1060,15 +1059,16 @@ public function journal() {
             $fieldsline['voucher_VatID']            = $row->VatID;
             $fieldsline['voucher_Vat']              = $row->Vat;
 
-            if(isset($this->headH['DepartmentID']) && $this->headH['DepartmentID'] > 0)
+            // If value(department/project) sent use that, othewise use defaults from product's accountplan (if set and enabled)
+            if(isset($this->headH['DepartmentID']))
                 $fieldsline['voucher_DepartmentID']     = $this->headH['DepartmentID'];
-            else
-                $fieldsline['voucher_DepartmentID']     = $productRow->CompanyDepartmentID;
+            elseif ($productAccountPlan->EnableDepartment == 1 && !is_null($productAccountPlan->DepartmentID))
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: `isset()` instead of `!is_null()`

File: modules/invoice/model/invoice.class.php [hunk 5]
```diff
@@ -1060,15 +1059,16 @@ public function journal() {
...
+                $fieldsline['voucher_DepartmentID']     = $productAccountPlan->DepartmentID;
 
-            if(isset($this->headH['ProjectID']) && $this->headH['ProjectID'] > 0)
+            if(isset($this->headH['ProjectID']))
                 $fieldsline['voucher_ProjectID']        = $this->headH['ProjectID'];
-            else
-                $fieldsline['voucher_ProjectID']        = $productRow->ProjectID;
+            elseif ($productAccountPlan->EnableProject == 1 && !is_null($productAccountPlan->ProjectID))
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: same as above

File: modules/invoicein/model/invoicein.class.php [hunk 6]
```diff
@@ -479,6 +477,26 @@ public function journal() {
...
+                        $VoucherH['voucher_AccountPlanID']  = $line->AccountPlanID;
+                        if($line->IsOnlyTax) {
+                            $VoucherH['voucher_AccountPlanID'] = 2710;
+                        }
+                        $line_accountplan = accounting::get_accountplan_object($VoucherH['voucher_AccountPlanID']);
+                        // set default car/department/project value
+                        $accountplan_default_car = NULL;
+                        $accountplan_default_department = NULL;
+                        $accountplan_default_project = NULL;
+                        if (!empty($line_accountplan) && $line_accountplan->EnableCar == 1 && !is_null($line_accountplan->CarID)) {
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php]: same as above

File: modules/invoicein/model/invoicein.class.php [hunk 7]
```diff
@@ -479,6 +477,26 @@ public function journal() {
...
+                          $accountplan_default_car = $line_accountplan->CarID;
+                        }
+                        if (!empty($line_accountplan) && $line_accountplan->EnableDepartment == 1 && !is_null($line_accountplan->DepartmentID)) {
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php]: same as above

File: modules/invoicein/model/invoicein.class.php [hunk 8]
```diff
@@ -479,6 +477,26 @@ public function journal() {
...
+                          $accountplan_default_department = $line_accountplan->DepartmentID;
+                        }
+                        if (!empty($line_accountplan) && $line_accountplan->EnableProject == 1 && !is_null($line_accountplan->ProjectID)) {
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php]: same as above

File: modules/invoicein/model/invoicein.class.php [hunk 9]
```diff
@@ -479,6 +477,26 @@ public function journal() {
                         $VoucherH['voucher_Description']    = '';
                         $VoucherH['voucher_AccountPlanID']  = 0;
 
+                        //#Motkonto resultat.
+                        $VoucherH['voucher_AccountPlanID']  = $line->AccountPlanID;
+                        if($line->IsOnlyTax) {
+                            $VoucherH['voucher_AccountPlanID'] = 2710;
+                        }
+                        $line_accountplan = accounting::get_accountplan_object($VoucherH['voucher_AccountPlanID']);
```
Comment (mladjo2505) [inline modules/invoicein/model/invoicein.class.php:485]: move to each separate else block so we don't do the setting of default if not necessary.

Comment (mladjo2505) [reply modules/invoicein/model/invoicein.class.php:485]: done

File: modules/journal/view/edit.php [hunk 10]
```diff
@@ -85,9 +85,9 @@
     $voucher_input->VoucherDate    = $voucherHead->VoucherDate     ? $voucherHead->VoucherDate     : $date->VoucherDate;
 }
 
-$voucher_input->CarID          = $voucherHead->CarID           ? $voucherHead->CarID           : $accountplan->CarID;
-$voucher_input->ProjectID      = $voucherHead->ProjectID       ? $voucherHead->ProjectID       : $accountplan->ProjectID;
-$voucher_input->DepartmentID   = $voucherHead->DepartmentID    ? $voucherHead->DepartmentID    : $accountplan->DepartmentID;
+if ($accountplan->EnableCar == 1 && isset($accountplan->CarID)) $voucher_input->CarID = $voucherHead->CarID ? $voucherHead->CarID : $accountplan->CarID;
```
Comment (mladjo2505) [inline modules/journal/view/edit.php]: test out

**Comment by mladjo2505** [comment]
Split into two commits since we might not want the invoice out logic, and that way it will be easier to discard.

**Comment by sls-lnluksa** [comment]
⚠️ 
To be deployed with FB#624

**Comment by mladjo2505** [comment]
After discussing with Arnt and Hildegunn this is finally done the way it should be, ready for final review.

**Review by stoivo — APPROVED** [review]

**Pull request merged**